### PR TITLE
Nim namespace for cpp

### PR DESCRIPTION
**Pull request by AdrianV**
I have created a new compiler option cppCompileToNamespace. When this option is enabled all generated nim code (without the main proc) is put into its own namespace Nim. This avoids possible name clashes when Nim code is compiled with other C++ code and Nim types have the same name as other types in the C++ code - for example common used names like TFrame or Exception.
There are also some small changes in order to make the compilation with the new C++ compiler from Embarcadero (formerly the Borland c++ compiler) possible. One was renaming the struct TFrame to TFrame_, but the typedef TFrame stays untouched. This was necessary because Embarcadero C++ freaks out when both have the same name out of some obscure reason - I guess it has to do with pre compiled headers, but I didn't investigate further. 
I know, that the bcc is not loved very much, but the new Embarcadero C++ is clang based and I need these changes, because I am working on a library which makes it possible to use the Borland VCL with Nim.

### CODE REVIEW

File: lib/nimbase.h [hunk 1]
```diff
@@ -425,26 +431,26 @@ struct TFrame {
 
 #ifdef NIM_NEW_MANGLING_RULES
   #define nimfr_(proc, file) \
-    TFrame FR_; \
```
Comment (Araq) [inline lib/nimbase.h]: Don't change the whitespace. We don't use tabulators.

**Comment by Araq** [comment]
Looks good, needs a changelog entry and documentation in ``advopt.txt``.

**Comment by dom96** [comment]
173+ merge commits?!? That doesn't look "good".

**Comment by dom96** [comment]
In any case, just please be sure to squash this before merging.

**Comment by AdrianV** [comment]
If you prefer I can create a new pull request which is based on current devel - this should be cleaner

**Comment by dom96** [comment]
It should be okay. Just need to make sure to squash this PR when merging it, otherwise our history will contains these hundreds of merges...

**Comment by dom96** [comment]
Now that this is merged, you should probably recreate your fork (remove and create it again) so that this doesn't happen again.

**Comment by AdrianV** [comment]
Thanks - I will remove my fork - I did that before I created my branch, but not locally.

**Pull request merged**